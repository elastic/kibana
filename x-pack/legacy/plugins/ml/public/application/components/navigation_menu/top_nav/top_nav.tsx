/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React, { FC, Fragment, useState, useEffect } from 'react';
import { Subscription } from 'rxjs';
import { EuiSuperDatePicker, OnRefreshProps } from '@elastic/eui';
import { TimeHistory } from 'ui/timefilter';
import { TimeRange } from 'src/plugins/data/public';

import {
  mlTimefilterRefresh$,
  mlTimefilterTimeChange$,
} from '../../../services/timefilter_refresh_service';
import { useUiContext } from '../../../contexts/ui/use_ui_context';
import { useUrlState } from '../../../util/url_state';

interface Duration {
  start: string;
  end: string;
}

function getRecentlyUsedRangesFactory(timeHistory: TimeHistory) {
  return function(): Duration[] {
    return (
      timeHistory.get()?.map(({ from, to }: TimeRange) => {
        return {
          start: from,
          end: to,
        };
      }) ?? []
    );
  };
}

function updateLastRefresh(timeRange: OnRefreshProps) {
  mlTimefilterRefresh$.next({ lastRefresh: Date.now(), timeRange });
}

export const TopNav: FC = () => {
  const { chrome, timefilter, timeHistory } = useUiContext();
  const [globalState, setGlobalState] = useUrlState('_g');
  const getRecentlyUsedRanges = getRecentlyUsedRangesFactory(timeHistory);

  const [refreshInterval, setRefreshInterval] = useState(
    globalState?.refreshInterval ?? timefilter.getRefreshInterval()
  );
  useEffect(() => {
    setGlobalState({ refreshInterval });
    timefilter.setRefreshInterval(refreshInterval);
  }, [refreshInterval?.pause, refreshInterval?.value]);

  const [time, setTime] = useState(timefilter.getTime());
  const [recentlyUsedRanges, setRecentlyUsedRanges] = useState(getRecentlyUsedRanges());
  const [isAutoRefreshSelectorEnabled, setIsAutoRefreshSelectorEnabled] = useState(
    timefilter.isAutoRefreshSelectorEnabled()
  );
  const [isTimeRangeSelectorEnabled, setIsTimeRangeSelectorEnabled] = useState(
    timefilter.isTimeRangeSelectorEnabled()
  );

  const dateFormat = chrome.getUiSettingsClient().get('dateFormat');

  useEffect(() => {
    const subscriptions = new Subscription();
    const refreshIntervalUpdate$ = timefilter.getRefreshIntervalUpdate$();
    if (refreshIntervalUpdate$ !== undefined) {
      subscriptions.add(refreshIntervalUpdate$.subscribe(timefilterUpdateListener));
    }
    const timeUpdate$ = timefilter.getTimeUpdate$();
    if (timeUpdate$ !== undefined) {
      subscriptions.add(timeUpdate$.subscribe(timefilterUpdateListener));
    }
    const enabledUpdated$ = timefilter.getEnabledUpdated$();
    if (enabledUpdated$ !== undefined) {
      subscriptions.add(enabledUpdated$.subscribe(timefilterUpdateListener));
    }

    return function cleanup() {
      subscriptions.unsubscribe();
    };
  }, []);

  useEffect(() => {
    // Force re-render with up-to-date values when isTimeRangeSelectorEnabled/isAutoRefreshSelectorEnabled are changed.
    timefilterUpdateListener();
  }, [isTimeRangeSelectorEnabled, isAutoRefreshSelectorEnabled]);

  function timefilterUpdateListener() {
    setTime(timefilter.getTime());
    setRefreshInterval(timefilter.getRefreshInterval());
    setIsAutoRefreshSelectorEnabled(timefilter.isAutoRefreshSelectorEnabled());
    setIsTimeRangeSelectorEnabled(timefilter.isTimeRangeSelectorEnabled());
  }

  function updateFilter({ start, end }: Duration) {
    const newTime = { from: start, to: end };
    // Update timefilter for controllers listening for changes
    timefilter.setTime(newTime);
    setTime(newTime);
    setRecentlyUsedRanges(getRecentlyUsedRanges());
    mlTimefilterTimeChange$.next({ lastRefresh: Date.now(), timeRange: { start, end } });
  }

  function updateInterval({
    isPaused: pause,
    refreshInterval: value,
  }: {
    isPaused: boolean;
    refreshInterval: number;
  }) {
    setRefreshInterval({ pause, value });
  }

  return (
    <Fragment>
      {(isAutoRefreshSelectorEnabled || isTimeRangeSelectorEnabled) && (
        <div className="mlNavigationMenu__topNav">
          <EuiSuperDatePicker
            start={time.from}
            end={time.to}
            isPaused={refreshInterval.pause}
            isAutoRefreshOnly={!isTimeRangeSelectorEnabled}
            refreshInterval={refreshInterval.value}
            onTimeChange={updateFilter}
            onRefresh={updateLastRefresh}
            onRefreshChange={updateInterval}
            recentlyUsedRanges={recentlyUsedRanges}
            dateFormat={dateFormat}
          />
        </div>
      )}
    </Fragment>
  );
};
