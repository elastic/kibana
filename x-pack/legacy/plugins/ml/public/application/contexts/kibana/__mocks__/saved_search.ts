/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

export const savedSearchMock: any = {
  id: 'the-saved-search-id',
  type: 'search',
  attributes: {
    title: 'the-saved-search-title',
    kibanaSavedObjectMeta: {
      searchSourceJSON:
        '{"highlightAll":true,"version":true,"query":{"query":"foo : \\"bar\\" ","language":"kuery"},"filter":[],"indexRefName":"kibanaSavedObjectMeta.searchSourceJSON.index"}',
    },
  },
  references: [
    {
      name: 'kibanaSavedObjectMeta.searchSourceJSON.index',
      type: 'index-pattern',
      id: 'the-index-pattern-id',
    },
  ],
  migrationVersion: { search: '7.5.0' },
  error: undefined,
};
