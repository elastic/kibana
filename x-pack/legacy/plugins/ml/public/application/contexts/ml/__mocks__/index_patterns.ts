/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { IndexPatternsContract } from '../../../../../../../../../src/plugins/data/public';

export const indexPatternsMock = (new (class {
  fieldFormats = [];
  config = {};
  savedObjectsClient = {};
  refreshSavedObjectsCache = {};
  clearCache = jest.fn();
  get = jest.fn();
  getDefault = jest.fn();
  getFields = jest.fn();
  getIds = jest.fn();
  getTitles = jest.fn();
  make = jest.fn();
})() as unknown) as IndexPatternsContract;
