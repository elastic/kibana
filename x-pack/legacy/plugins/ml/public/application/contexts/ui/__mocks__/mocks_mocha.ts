/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { Subject } from 'rxjs';

export const uiChromeMock = {
  getBasePath: () => 'basePath',
  getUiSettingsClient: () => {
    return {
      get: (key: string) => {
        switch (key) {
          case 'dateFormat':
            return 'MMM D, YYYY @ HH:mm:ss.SSS';
          case 'theme:darkMode':
            return false;
          case 'timepicker:timeDefaults':
            return {};
          case 'timepicker:refreshIntervalDefaults':
            return { pause: false, value: 0 };
          default:
            throw new Error(`Unexpected config key: ${key}`);
        }
      },
    };
  },
};

interface RefreshInterval {
  value: number;
  pause: boolean;
}

const time = {
  from: 'Thu Aug 29 2019 02:04:19 GMT+0200',
  to: 'Sun Sep 29 2019 01:45:36 GMT+0200',
};

export const uiTimefilterMock = {
  isAutoRefreshSelectorEnabled() {
    return this._isAutoRefreshSelectorEnabled;
  },
  isTimeRangeSelectorEnabled() {
    return this._isTimeRangeSelectorEnabled;
  },
  enableAutoRefreshSelector() {
    this._isAutoRefreshSelectorEnabled = true;
  },
  enableTimeRangeSelector() {
    this._isTimeRangeSelectorEnabled = true;
  },
  getActiveBounds() {
    return;
  },
  getEnabledUpdated$() {
    return { subscribe: () => {} };
  },
  getFetch$() {
    return new Subject();
  },
  getRefreshInterval() {
    return this.refreshInterval;
  },
  getRefreshIntervalUpdate$() {
    return { subscribe: () => {} };
  },
  getTime: () => time,
  getTimeUpdate$() {
    return { subscribe: () => {} };
  },
  _isAutoRefreshSelectorEnabled: false,
  _isTimeRangeSelectorEnabled: false,
  refreshInterval: { value: 0, pause: true },
  on: (event: string, reload: () => void) => {},
  setRefreshInterval(refreshInterval: RefreshInterval) {
    this.refreshInterval = refreshInterval;
  },
};

export const uiTimeHistoryMock = {
  get: () => [time],
};
