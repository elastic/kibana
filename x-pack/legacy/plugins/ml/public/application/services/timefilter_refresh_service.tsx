/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { Subject } from 'rxjs';

interface MlTimeFilterUpdate {
  lastRefresh: number;
  timeRange: { start: string; end: string };
}

export const mlTimefilterRefresh$ = new Subject<MlTimeFilterUpdate>();
export const mlTimefilterTimeChange$ = new Subject<MlTimeFilterUpdate>();
