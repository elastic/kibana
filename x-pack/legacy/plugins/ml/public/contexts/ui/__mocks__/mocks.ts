/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

export const uiChromeMock = {
  getBasePath: () => 'basePath',
  getUiSettingsClient: () => {
    return {
      get: (key: string) => {
        switch (key) {
          case 'dateFormat':
          case 'timepicker:timeDefaults':
            return {};
          case 'timepicker:refreshIntervalDefaults':
            return { pause: false, value: 0 };
          default:
            throw new Error(`Unexpected config key: ${key}`);
        }
      },
    };
  },
};

export const uiTimefilterMock = {
  getRefreshInterval: () => '30s',
  getTime: () => ({ from: 0, to: 0 }),
  on: (event: string, reload: () => void) => {},
};

export const uiTimeHistoryMock = {
  get: () => [{ from: 0, to: 0 }],
};
