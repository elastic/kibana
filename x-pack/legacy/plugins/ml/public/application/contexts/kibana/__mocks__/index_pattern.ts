/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { IndexPattern } from 'ui/index_patterns';

export const indexPatternMock = ({
  id: 'the-index-pattern-id',
  title: 'the-index-pattern-title',
  fields: [],
} as unknown) as IndexPattern;
