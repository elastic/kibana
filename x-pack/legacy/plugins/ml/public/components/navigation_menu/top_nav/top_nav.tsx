/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React, { FC, Fragment, useState, useEffect } from 'react';
import { EuiSuperDatePicker } from '@elastic/eui';
import { TimeHistory, TimeRange } from 'ui/timefilter/time_history';

import { mlTimefilterRefresh$ } from '../../../services/timefilter_refresh_service';
import { useUiContext } from '../../../contexts/ui/use_ui_context';

interface Duration {
  start: string;
  end: string;
}

function getRecentlyUsedRangesFactory(timeHistory: TimeHistory) {
  return function(): Duration[] {
    return timeHistory.get().map(({ from, to }: TimeRange) => {
      return {
        start: from,
        end: to,
      };
    });
  };
}

export const TopNav: FC = () => {
  const { chrome, timefilter, timeHistory } = useUiContext();
  const getRecentlyUsedRanges = getRecentlyUsedRangesFactory(timeHistory);

  const [refreshInterval, setRefreshInterval] = useState(timefilter.getRefreshInterval());
  const [time, setTime] = useState(timefilter.getTime());
  const [recentlyUsedRanges, setRecentlyUsedRanges] = useState(getRecentlyUsedRanges());
  const [isAutoRefreshSelectorEnabled, setIsAutoRefreshSelectorEnabled] = useState(
    timefilter.isAutoRefreshSelectorEnabled
  );
  const [isTimeRangeSelectorEnabled, setIsTimeRangeSelectorEnabled] = useState(
    timefilter.isTimeRangeSelectorEnabled
  );

  const dateFormat = chrome.getUiSettingsClient().get('dateFormat');

  useEffect(() => {
    timefilter.on('refreshIntervalUpdate', timefilterUpdateListener);
    timefilter.on('timeUpdate', timefilterUpdateListener);
    timefilter.on('enabledUpdated', timefilterUpdateListener);

    return function cleanup() {
      timefilter.off('refreshIntervalUpdate', timefilterUpdateListener);
      timefilter.off('timeUpdate', timefilterUpdateListener);
      timefilter.off('enabledUpdated', timefilterUpdateListener);
    };
  }, []);

  useEffect(() => {
    // Force re-render with up-to-date values when isTimeRangeSelectorEnabled/isAutoRefreshSelectorEnabled are changed.
    timefilterUpdateListener();
  }, [isTimeRangeSelectorEnabled, isAutoRefreshSelectorEnabled]);

  function timefilterUpdateListener() {
    setTime(timefilter.getTime());
    setRefreshInterval(timefilter.getRefreshInterval());
    setIsAutoRefreshSelectorEnabled(timefilter.isAutoRefreshSelectorEnabled);
    setIsTimeRangeSelectorEnabled(timefilter.isTimeRangeSelectorEnabled);
  }

  function updateFilter({ start, end }: Duration) {
    const newTime = { from: start, to: end };
    // Update timefilter for controllers listening for changes
    timefilter.setTime(newTime);
    setTime(newTime);
    setRecentlyUsedRanges(getRecentlyUsedRanges());
  }

  function updateInterval({
    isPaused,
    refreshInterval: interval,
  }: {
    isPaused: boolean;
    refreshInterval: number;
  }) {
    const newInterval = {
      pause: isPaused,
      value: interval,
    };
    // Update timefilter for controllers listening for changes
    timefilter.setRefreshInterval(newInterval);
    // Update state
    setRefreshInterval(newInterval);
  }

  return (
    <Fragment>
      {(isAutoRefreshSelectorEnabled || isTimeRangeSelectorEnabled) && (
        <div className="mlNavigationMenu__topNav">
          <EuiSuperDatePicker
            start={time.from}
            end={time.to}
            isPaused={refreshInterval.pause}
            isAutoRefreshOnly={!isTimeRangeSelectorEnabled}
            refreshInterval={refreshInterval.value}
            onTimeChange={updateFilter}
            onRefresh={() => mlTimefilterRefresh$.next()}
            onRefreshChange={updateInterval}
            recentlyUsedRanges={recentlyUsedRanges}
            dateFormat={dateFormat}
          />
        </div>
      )}
    </Fragment>
  );
};
