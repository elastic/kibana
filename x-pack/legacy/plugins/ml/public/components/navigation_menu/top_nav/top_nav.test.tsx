/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { mount, shallow } from 'enzyme';
import React from 'react';

import { uiTimefilterMock } from '../../../contexts/ui/__mocks__/mocks';
import { mlTimefilterRefresh$ } from '../../../services/timefilter_refresh_service';

import { MlSuperDatePickerWithUpdate, TopNav } from './top_nav';

uiTimefilterMock.enableAutoRefreshSelector();
uiTimefilterMock.enableTimeRangeSelector();

jest.mock('../../../contexts/ui/use_ui_context');

const noop = () => {};

describe('Navigation Menu: <TopNav />', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('Minimal initialization.', () => {
    const refreshListener = jest.fn();
    const refreshSubscription = mlTimefilterRefresh$.subscribe(refreshListener);

    const wrapper = shallow(<TopNav />);
    expect(wrapper).toMatchSnapshot();
    expect(refreshListener).toBeCalledTimes(0);

    refreshSubscription.unsubscribe();
  });

  // The following tests are written against MlSuperDatePickerWithUpdate
  // instead of TopNav. TopNav uses hooks and we cannot writing tests
  // with async hook updates yet until React 16.9 is available.

  // MlSuperDatePickerWithUpdate fixes an issue with EuiSuperDatePicker
  // which didn't make it into Kibana 7.4. We should be able to just
  // use EuiSuperDatePicker again once the following PR is in EUI:
  // https://github.com/elastic/eui/pull/2298

  test('Listen for consecutive super date picker refreshs.', async () => {
    const onRefresh = jest.fn();

    const componentRefresh = mount(
      <MlSuperDatePickerWithUpdate
        onTimeChange={noop}
        isPaused={false}
        onRefresh={onRefresh}
        refreshInterval={10}
      />
    );

    const instanceRefresh = componentRefresh.instance();

    jest.advanceTimersByTime(10);
    // @ts-ignore
    await instanceRefresh.asyncInterval.__pendingFn;
    jest.advanceTimersByTime(10);
    // @ts-ignore
    await instanceRefresh.asyncInterval.__pendingFn;

    expect(onRefresh).toBeCalledTimes(2);
  });

  test('Switching refresh interval to pause should stop onRefresh being called.', async () => {
    const onRefresh = jest.fn();

    const componentRefresh = mount(
      <MlSuperDatePickerWithUpdate
        onTimeChange={noop}
        isPaused={false}
        onRefresh={onRefresh}
        refreshInterval={10}
      />
    );

    const instanceRefresh = componentRefresh.instance();

    jest.advanceTimersByTime(10);
    // @ts-ignore
    await instanceRefresh.asyncInterval.__pendingFn;
    componentRefresh.setProps({ isPaused: true, refreshInterval: 0 });
    jest.advanceTimersByTime(10);
    // @ts-ignore
    await instanceRefresh.asyncInterval.__pendingFn;

    expect(onRefresh).toBeCalledTimes(1);
  });
});
