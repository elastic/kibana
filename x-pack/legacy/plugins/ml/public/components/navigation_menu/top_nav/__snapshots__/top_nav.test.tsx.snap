// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Menu: <TopNav /> Minimal initialization. 1`] = `
<Fragment>
  <div
    className="mlNavigationMenu__topNav"
  >
    <EuiSuperDatePicker
      commonlyUsedRanges={
        Array [
          Object {
            "end": "now/d",
            "label": "Today",
            "start": "now/d",
          },
          Object {
            "end": "now/w",
            "label": "This week",
            "start": "now/w",
          },
          Object {
            "end": "now/M",
            "label": "This month",
            "start": "now/M",
          },
          Object {
            "end": "now/y",
            "label": "This year",
            "start": "now/y",
          },
          Object {
            "end": "now-1d/d",
            "label": "Yesterday",
            "start": "now-1d/d",
          },
          Object {
            "end": "now",
            "label": "Week to date",
            "start": "now/w",
          },
          Object {
            "end": "now",
            "label": "Month to date",
            "start": "now/M",
          },
          Object {
            "end": "now",
            "label": "Year to date",
            "start": "now/y",
          },
        ]
      }
      dateFormat="MMM D, YYYY @ HH:mm:ss.SSS"
      end="Sun Sep 29 2019 01:45:36 GMT+0200"
      isAutoRefreshOnly={false}
      isDisabled={false}
      isPaused={true}
      onRefresh={[Function]}
      onRefreshChange={[Function]}
      onTimeChange={[Function]}
      recentlyUsedRanges={
        Array [
          Object {
            "end": "Sun Sep 29 2019 01:45:36 GMT+0200",
            "start": "Thu Aug 29 2019 02:04:19 GMT+0200",
          },
        ]
      }
      refreshInterval={0}
      showUpdateButton={true}
      start="Thu Aug 29 2019 02:04:19 GMT+0200"
    />
  </div>
</Fragment>
`;
